src/ParseObject.js

```diff
@@ -172,22 +177,26 @@ export default class ParseObject {
   }
 
   /**
-   * Returns a local or server Id used to pull data from the Object State store
-   * If single instance objects are disabled, it will use the object's unique
-   * count to separate its data from other objects with the same server Id.
+   * Returns a unique identifier used to pull data from the State Controller.
    */
   _getStateIdentifier(): string {
-    if (typeof this.id === 'string') {
-      if (singleInstance) {
-        return this.id;
+    if (singleInstance) {
+      let id = this.id;
+      if (!id) {
+        id = this._getId();
       }
-      return this.id + '_' + String(this._objCount);
+      return {
```

Comment:
... which can be `{ id: string, className: string}`, and not always `ParseObject`.